File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include<bits/stdc++.h>
using namespace std;

bool visited[103][103];
int arr[103][103];
int dx[] = { 0,0,-1,1 };
int dy[] = { 1,-1,0,0 };
int len, T;

void DFS(int y, int x) {
	for (int i = 0; i < 4; i++) {

		int nx = x + dx[i];
		int ny = y + dy[i];

		if (nx < 0 || nx >= T || ny < 0 || ny >= T || visited[ny][nx]) continue;
		if (arr[ny][nx] > len) {
			visited[ny][nx] = 1;
			DFS(ny, nx);
		}
	}
}


int main() {
	int result;
	int maxResult = 0;
	int maxNum = 0;

	cin >> T;
	for (int i = 0; i < T; i++) {
		for (int j = 0; j < T; j++) {
			cin >> arr[i][j];
			maxNum = max(maxNum, arr[i][j]);
		}
	}
	for (len = 0; len <= maxNum; len++) {
		memset(visited, 0, sizeof(visited));
		result = 0;
		for (int i = 0; i < T; i++) {
			for (int j = 0; j < T; j++) {
				if (len < arr[i][j] && !visited[i][j]) {
					visited[i][j] = 1;
					DFS(i, j);
					result++;
				}
			}
		}
		maxResult = max(maxResult, result);
	}
	cout << maxResult;
	

	return 0;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.util.*;

public class Main {
    static final int maxSize = 104;
    static int[][] arr = new int[maxSize][maxSize];
    static int[][] visited = new int[maxSize][maxSize];
    static int N;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int maxH = 0;
        int rnt = 0;

        N = sc.nextInt();
        for (int y = 0; y < N; y++) {
            for (int x = 0; x < N; x++) {
                arr[y][x] = sc.nextInt();
                maxH = Math.max(maxH, arr[y][x]);
            }
        }

        for (int height = 0; height <= maxH; height++) {
            for (int[] row : visited) {
                Arrays.fill(row, 0);
            }

            int cnt = 0;
            for (int y = 0; y < N; y++) {
                for (int x = 0; x < N; x++) {
                    if (arr[y][x] > height && visited[y][x] == 0) {
                        cnt++;
                        visited[y][x] = 1;
                        DFS(y, x, height);
                    }
                }
            }
            rnt = Math.max(rnt, cnt);
        }

        System.out.println(rnt);
    }

    static void DFS(int y, int x, int height) {
        for (int k = 0; k < 4; k++) {
            int ny = y + dy[k];
            int nx = x + dx[k];
            if (ny < 0 || nx < 0 || ny >= N || nx >= N || visited[ny][nx] == 1) continue;
            if (arr[ny][nx] <= height) continue;
            visited[ny][nx] = 1;
            DFS(ny, nx, height);
        }
    }
}